spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/PredictionService.scala

```diff
@@ -0,0 +1,258 @@
+ * limitations under the License.
+ */
+
+package com.intel.analytics.bigdl.optim
+
+import java.util.concurrent.{Executors, LinkedBlockingQueue}
+
+import com.intel.analytics.bigdl.Module
+import com.intel.analytics.bigdl.nn.abstractnn.Activity
+import com.intel.analytics.bigdl.serialization.Bigdl.AttrValue.ArrayValue
+import com.intel.analytics.bigdl.serialization.Bigdl.{AttrValue, BigDLTensor, DataType, TensorStorage}
+import com.intel.analytics.bigdl.tensor.Tensor
+import com.intel.analytics.bigdl.tensor.TensorNumericMath.TensorNumeric
+import com.intel.analytics.bigdl.tensor.TensorNumericMath.TensorNumeric.{NumericBoolean, NumericChar, NumericDouble, NumericFloat, NumericInt, NumericLong, NumericString}
+import com.intel.analytics.bigdl.utils.serializer.converters.DataConverter
+import com.intel.analytics.bigdl.utils.serializer.{DeserializeContext, ModuleSerializer, ProtoStorageType, SerializeContext}
+import com.intel.analytics.bigdl.utils.{T, Table}
+
+import scala.collection.JavaConverters._
+import scala.collection.mutable
+import scala.concurrent.{ExecutionContext, ExecutionContextExecutor, Future}
+import scala.reflect.ClassTag
+import scala.reflect.runtime.universe.Type
+
+/**
+ * <h6>Prediction Service Interface for Concurrent Calls</h6>
+ * In this service, several `model instances` sharing weights/bias will be built.
+ * <br>
+ * A `BlockingQueue` will be built to maintain available `model instances`.
+ * <br>
+ * When `predict` method called, service will try to take an instance from `BlockingQueue`,
+ * which means if all instances are on serving, the predicting request will be blocked until
+ * some instances are released.
+ * <br>
+ * The release method will be executed asynchronously when predictions finished.
+ *
+ * @param model BigDL model used to do predictions
+ * @param nInstances number of model instances
+ * @param nBackendThreads number of threads to release instances after predictions.
+ */
+class PredictionService[T: ClassTag] private[optim](
+  model: Module[T],
+  nInstances: Int = 10,
+  nBackendThreads: Int = 2
+)(implicit ev: TensorNumeric[T]) {
+
+  require(nInstances > 1, "At least two instances are needed!")
+
+  require(nBackendThreads >= 1, "At least one backend threads are needed!")
+
+  // ThreadPool for scala.concurrent.Future,
+  // which is build for keeping threads available for unlocking to avoid dead locks.
+  private implicit val ec: ExecutionContextExecutor = {
+    ExecutionContext.fromExecutor(Executors.newFixedThreadPool(nBackendThreads))
+  }
+
+  protected val instQueue: LinkedBlockingQueue[Module[T]] = {
+    val shallowCopies = (1 to nInstances)
+      .map(_ => model.clone(false).evaluate()).asJava
+
+    new LinkedBlockingQueue[Module[T]](shallowCopies)
+  }
+
+  def predict(request: Activity): Activity = {
+    val module = fetchInstance()
+    // cloned values after prediction finished
+    val output = module.forward(request) match {
```

Comment:
forward may throw exception if the request is illegal, please catch the exception and releaseInstance anyway.

Reply:
code pieces added to catch exceptions